### PR TITLE
Reorganise the rich build operation details/results types to further decouple from producer side semantics

### PR DESCRIPTION
**Pull request by alkemist**
The goal here is to increase our chances of maintaining backwards compatibility, by making the types that are shared between the core and the scan plugin as light as possible.

### CODE REVIEW

File: subprojects/base-services/src/main/java/org/gradle/internal/operations/BuildOperationType.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * outside of listener callback to process it off thread.
+ * It can be assumed that the objects are not held longer than is necessary to process/transform them
+ * into detached representations suitable for the build scan event stream.
+ *
+ * @param <D> the type of details object for the operation
+ * @param <R> the type of result object for the operation
+ * @since 4.0
+ */
+@SuppressWarnings("unused")
+public interface BuildOperationType<D, R> {
```
Comment (oehme) [inline subprojects/base-services/src/main/java/org/gradle/internal/operations/BuildOperationType.java]: I don't understand what benefit this brings over `BuildOperationDetails<RESULT>`.

File: subprojects/core/src/main/java/org/gradle/api/execution/internal/ExecuteTaskBuildOperation.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+package org.gradle.api.execution.internal;
+
+import org.gradle.api.Task;
+import org.gradle.api.internal.TaskInternal;
+import org.gradle.internal.operations.BuildOperationType;
+import org.gradle.internal.scan.UsedByScanPlugin;
+
+public final class ExecuteTaskBuildOperation implements BuildOperationType<ExecuteTaskBuildOperation.Details, Void> {
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/api/execution/internal/ExecuteTaskBuildOperation.java]: All of these should be renamed to xxxBuildOperationType (if we want to go with that pattern). They are not `BuildOperation`s (a thing that gets executed), but metadata describing that operation.

File: subprojects/core/src/main/java/org/gradle/api/execution/internal/ExecuteTaskBuildOperation.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    @UsedByScanPlugin
+    public interface Details {
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/api/execution/internal/ExecuteTaskBuildOperation.java]: What benefit does an interface bring here? All of these types should be immutable value types.

File: subprojects/base-services/src/main/java/org/gradle/internal/operations/BuildOperationType.java (hunk 1 above)
Comment (alkemist) [inline subprojects/base-services/src/main/java/org/gradle/internal/operations/BuildOperationType.java]: It avoids build scans having to link against any infrastructure types, which means that we don't have to commit to how this stuff is packaged right now.

**Comment by alkemist** [comment]
I have already asked @eriwen to rename the existing `BuildOperationType` enum. He agreed and will do that on Monday.

**Comment by alkemist** [comment]
> Some insight into the motivation behind each of the added indirections

It's about making it easier to achieve backwards compatibility in two ways

1. Not hampering the evolution of the producer side of build operations by requiring it to be stable
2. Minimising the static reach from types that cross the core/scan-plugin boundary in order to minimise accidental leakage of unstable types

**Comment by alkemist** [comment]
> now need to write 4-6 types for each build operation

Only for the operations that effectively expose a stable, backwards compatible, API. We are paying a price for that stability in this ceremony. However, I am certain that without this ceremony, we will not be able to keep these types stable.

Counting types is a misleading metric. There's one container type with no behaviour, then two contract interfaces (which are absolutely a requirement), then two implementations. We _could_ do away with the container type and have no way of enforcing compile time assistance for the producer side API, but that seems a might high cost to pay to avoid a few characters.

File: subprojects/core/src/main/java/org/gradle/api/execution/internal/ExecuteTaskBuildOperation.java (hunk 3 above)
Comment (alkemist) [inline subprojects/core/src/main/java/org/gradle/api/execution/internal/ExecuteTaskBuildOperation.java]: This was effectively a requirement from @adammurdoch. It leaves the door more open to the consumer providing the type and having the producer adapt, like the TAPI does. 

I'm ok with punting on this and allowing classes to cross the boundary.

File: subprojects/core/src/main/java/org/gradle/api/execution/internal/ExecuteTaskBuildOperation.java (hunk 2 above)
Comment (alkemist) [inline subprojects/core/src/main/java/org/gradle/api/execution/internal/ExecuteTaskBuildOperation.java]: I've made this change. I initially considered it redundant, but I don't object to the convention.

**Comment by oehme** [comment]
> Only for the operations that effectively expose a stable, backwards compatible, API.

This part convinced me :) As long as the API still allows us to quickly add "simple" operations with no contract towards build scans, we're good.

**Pull request closed without merging**